### ISSUE TITLE
Sparsity regulariser not implemented correctly?

### ISSUE THREAD

**Issue by reynoldscem** [yashbhalgat/HashNeRF-pytorch#14]
[Here](https://github.com/yashbhalgat/HashNeRF-pytorch/blob/a7d64eeb8844b57f5ba90463185a1506e2cbb4b8/run_nerf.py#L357) you use the weights of the ray to compute entropy. These weights however are un-normalized, so the calculation of entropy isn't right. I think that either normalizing the weights, or just passing them as `logits` rather than `probs` should be correct.

**Comment by Feynman1999** [comment]
same question? and have you tried the sparse loss?

**Comment by Feynman1999** [comment]
> [Here](https://github.com/yashbhalgat/HashNeRF-pytorch/blob/a7d64eeb8844b57f5ba90463185a1506e2cbb4b8/run_nerf.py#L357) you use the weights of the ray to compute entropy. These weights however are un-normalized, so the calculation of entropy isn't right. I think that either normalizing the weights, or just passing them as `logits` rather than `probs` should be correct.

update: I found that in the init method of the torch Categorical, there is a normalization:
self.probs = probs / probs.sum(-1, keepdim=True)

link: https://pytorch.org/docs/stable/_modules/torch/distributions/categorical.html#Categorical

**Comment by jjparkcv** [comment]
Can someone provide a reference to the sparsity reguarlization?

**Comment by reynoldscem** [comment]
@jjparkcv in the code or a paper?

**Comment by jjparkcv** [comment]
@reynoldscem I meant a paper. Sorry for the confusion.

**Comment by reynoldscem** [comment]
@jjparkcv No problem! [This regulariser is in this paper](https://openaccess.thecvf.com/content/CVPR2022/html/Kim_InfoNeRF_Ray_Entropy_Minimization_for_Few-Shot_Neural_Volume_Rendering_CVPR_2022_paper.html), but it may also appear elsewhere! It strikes me as the sort of thing that might have been done a couple / a few times independently.

**Comment by yashbhalgat** [comment]
@reynoldscem thanks for pointing this out! And apologies for the delayed response on this issue. The bug is now fixed in the latest commit [041e36f](https://github.com/yashbhalgat/HashNeRF-pytorch/commit/041e36fd8a5891a18775ab75b4e8c48b2867c3ab).

The weights vector should be normalized as: `[w_1, w_2, ..., w_n, 1-sum(w_1, ..., w_n)]`.
On this [line](https://github.com/yashbhalgat/HashNeRF-pytorch/blob/041e36fd8a5891a18775ab75b4e8c48b2867c3ab/run_nerf.py#L364), you will find a weird constant `1e-6`. That's because floating point precision sometimes causes `sum(w_1, ..., w_n)` to exceed 1.

Closing this issue now. Feel free to reopen if needed!

**Comment by reynoldscem** [comment]
@yashbhalgat Thanks, that's perfect
